homeassistant/components/stream/core.py

```diff
@@ -42,17 +64,69 @@ class Segment:
     duration: float = attr.ib(default=0)
     # For detecting discontinuities across stream restarts
     stream_id: int = attr.ib(default=0)
-    parts: list[Part] = attr.ib(factory=list)
+    # Parts are stored in a dict indexed by http range start for easy
+    # lookup when requested by byterange
+    # As of Python 3.7, insertion order is preserved, and we insert
+    # in sequential order, so the Parts are ordered
+    parts_by_http_range: dict[int, Part] = attr.ib(factory=dict)
```

Comment:
> 
> 
> I had a couple suggestions:
> 
>     * My impression is that this dict is a map of the part.data where the key is the size of all previous part.data bytes (all the part.data without the init).   It might be somewhat easier to follow if variable was named in terms of part and data only, without mentioning http range here.  My impression is the http range maps directly to the part.data values, and so it can keep the concept of an http range limited to the hls code.
> 

How about just keeping it named parts?

>     * Perhaps moving the "part_put" method to `Segment` may be useful.  It could accept a `Part`, and could hide updating the range map inside `Segment` -- could even consider updating `duration` too on the last part.  That method would probably need to update the start of the map based on the last part.data size, rather than in terms of _memory_file position subtracting out index, but i think that might be a good thing to more easily see what this index is doing.

We can do that, but then should we consider moving put to Segment as well?

Reply:
Renamed the variables as per your suggestions and used `parts_by_byterange` for the parts, removing http from the name.

Reply:
>     * Perhaps moving the "part_put" method to `Segment` may be useful.  It could accept a `Part`, and could hide updating the range map inside `Segment` -- could even consider updating `duration` too on the last part.  That method would probably need to update the start of the map based on the last part.data size, rather than in terms of _memory_file position subtracting out index, but i think that might be a good thing to more easily see what this index is doing.

`Segment` doesn't have the outputs, so we'd have to pass those in somehow. Moving the part_event into `Segment` is worse.

Reply:
Moved the calls for `put` and `part_put` into `Segment` methods. Had to pass in the outputs every time we create the `Segment`, which makes sense since we were already calling the callback each time anyway. The way I did it means the callback only gets called once per Segment which I think should be fine.

Reply:
Ended up calling the new `Segment.add_part` method from the main thread event loop. This keeps all the dict accesses on the main thread to help avoid any threading issues, and we can also avoid copying the Segment / dict during the playlist render.